app/lib/exchange-rate/money.ts

```diff
@@ -0,0 +1,480 @@
+            ...options,
+            decimals: this.decimals,
+          });
+          formattedValue = formattedValue.replace('BTC', '').trim();
+          return `${this.symbol}${formattedValue}`;
+        },
+      },
+      {
+        name: 'satoshi',
+        decimals: 0,
+        symbol: '₿',
+        factor: new Big(10 ** -8),
+        format: function (value: number, options: FormatOptions = {}) {
+          let formattedValue = baseFormat(value, {
+            ...options,
+            decimals: this.decimals,
+          });
+          formattedValue = formattedValue.replace('BTC', '').trim();
+          return `${formattedValue}${this.symbol}`;
+        },
+      },
+      {
+        name: 'millisatoshi',
+        decimals: 0,
+        symbol: 'msat',
+        factor: new Big(10 ** -11),
+        format: function (value: number, options: FormatOptions = {}) {
+          let formattedValue = baseFormat(value, {
+            ...options,
+            decimals: this.decimals,
+          });
+          formattedValue = formattedValue.replace('BTC', '').trim();
+          return `${formattedValue} ${this.symbol}`;
+        },
+      },
+    ],
+  },
+};
+
+type MoneyData = {
+  /**
+   * Currency of the money
+   */
+  currency: string;
+  /**
+   * Amount of the money stored in the `unit` format
+   */
+  amount: Big;
+  /**
+   * Unit in which the `amount` is stored. It is always the minimal supported unit for the `currency` (e.g. millisatoshi
+   * for BTC)
+   */
+  amountUnit: CurrencyUnit;
+  /**
+   * Unit of the initial amount that was provided when creating the money
+   */
+  initialUnit: CurrencyUnit;
+};
+
+export class Money {
+  private readonly _data: MoneyData;
+
+  constructor(data: MoneyInput) {
+    const { baseUnit, minUnit, selectedUnit } =
+      Money.getCurrencyDataForInput(data);
+    const unit = selectedUnit ?? baseUnit;
+
+    const multiplier = unit.factor.div(minUnit.factor);
+    const amountInMinUnit = new Big(data.amount).mul(multiplier);
+
+    this._data = {
+      amount: amountInMinUnit.round(minUnit.decimals, Big.roundDown),
+      currency: data.currency,
+      amountUnit: minUnit,
+      initialUnit: unit,
+    };
+
+    Object.freeze(this);
+    Object.freeze(this._data);
+  }
+
+  static create(props: MoneyInput): Money {
+    return new Money(props);
+  }
+
+  /**
+   * Sum an array of moneys.
+   *
+   * If the array is empty, a currency must be specified so that 0 can be returned in that currency.
+   *
+   * The precision, etc., is based on the first item in the array.
+   * If the array is empty, the options object will be used instead.
+   */
+  static sum(moneys: Money[], currency?: string): Money {
+    if (moneys.length === 0 && currency === undefined) {
+      throw new Error(
+        "Currency must be set when summing an empty list of money's",
+      );
+    }
+
+    if (moneys.length === 0 && currency !== undefined) {
+      return new Money({ amount: 0, currency });
```

Comment:
I think yeah that is the idea when having static factory methods. then you can even make constructor private (that's when I normally use this pattern where there is some additional logic in the factory method that you don't want in the constructor so you put it there and make constructor private) but in this exact case it's more because I copied the approach from the lib and didn't change that part.

we can maybe even lose the static method and just use constructor everywhere? or you would prefer to keep it?

Reply:
I don't have a preference, but might as well just use the constructor everywhere

Reply:
I only have a slight style preference. With constructors the chaining looks a bit awkward:
```
new Money({ ... }).convert(...)
```
vs
```
Money.create({ ... }).convert(...)
```

though Money.create is longer to write. Money.of would be shorter but I'll just remove it and leave constuctor for now